### PR TITLE
feat(KLV-30): add mlflow support for model registry

### PR DESCRIPTION
**Pull request by Thrimbda**
**What this PR does / why we need it**:

Add your description

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #
add mlflow model support

Reference to #

**Special notes for your reviewer**:

/cc @your-reviewer

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: scripts/serving/wrapper/preprocessor.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 SKLEARN_MODEL = "model.joblib"
 XGBOOST_MODEL = "model.xgboost"
 
+# FIXME: why do we even need this class?
```
Comment (simon-cj) [inline scripts/serving/wrapper/preprocessor.py:16]: What's meaning? We can discuss it， if don‘t need it， we can create a issue to fix it， not  suggest comment in code

File: pkg/apis/modeljob/v1alpha1/modeljob_types.go [hunk 2]
```diff
@@ -131,6 +131,8 @@ const (
 	FormatTensorRT    Format = "TensorRT"
 	FormatSKLearn     Format = "SKLearn"
 	FormatXGBoost     Format = "XGBoost"
+	// TODO: we have exactly same data structure in ormb
```
Comment (simon-cj) [inline pkg/apis/modeljob/v1alpha1/modeljob_types.go:134]: Create issue to fix it

**Comment by simon-cj** [comment]
PR 提交的标题按照 https://github.com/kleveross/klever-model-registry/blob/master/CONTRIBUTING.md 参考这个规范

File: scripts/serving/wrapper/preprocessor.py (hunk 1 above)
Comment (Thrimbda) [reply scripts/serving/wrapper/preprocessor.py:16]: make sense

Comment (Thrimbda) [reply scripts/serving/wrapper/preprocessor.py:16]: I'll remove these comments once finished, please note this PR is still work in progress, thanks for noticing

**Comment by Thrimbda** [comment]
> PR 提交的标题按照 https://github.com/kleveross/klever-model-registry/blob/master/CONTRIBUTING.md 参考这个规范

明白，不过目前确实是WIP

File: pkg/registry/serving/composer.go [hunk 3]
```diff
@@ -337,14 +350,18 @@ func composeInitContainer(sdep *seldonv1.SeldonDeployment) error {
...
+			// https://github.com/SeldonIO/seldon-core/blob/0ef45fd234a674fc9b6c8d034cd2e42b4c9ebd05/operator/controllers/model_initializer_injector.go#L118
+			Name:  p.Name + "-" + "model-initializer",
 			Image: viper.GetString(envModelInitializerImage),
 			Args:  []string{p.Graph.ModelURI, modelMountPath},
 			// Get username and password from environment
 			// Here AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID are used
 			// because Seldon Core does not support renaming the environment variable name.
 			// it is used in ormb-storage-initializer.
-			// please refenence https://github.com/kleveross/ormb/blob/master/cmd/ormb-storage-initializer/cmd/pull-and-export.go#L47
+			// please refenence https://github.com/kleveross/ormb/blob/4ed110bdcd63540b7feaa179ebed2cb94c2aaba2/cmd/ormb-storage-initializer/cmd/pull-and-export.go#L47
```
Comment (simon-cj) [inline pkg/registry/serving/composer.go:364]: Why we need replace `master` with `4ed110bdcd63540b7feaa179ebed2cb94c2aaba2`

File: pkg/registry/serving/composer.go [hunk 4]
```diff
@@ -79,33 +79,39 @@ func validateComponentSpecs(p *seldonv1.PredictorSpec) error {
...
-		}
+		setupNoEngineMode(&p)
 	}
 
 	composeInitContainer(sdep)
 
 	return nil
 }
 
+func setupNoEngineMode(pu *seldonv1.PredictorSpec) {
```
Comment (simon-cj) [inline pkg/registry/serving/composer.go:105]: setupNoEngineMode should set no engine mode, should not set pu.Name, WDYT

File: pkg/registry/serving/composer.go (hunk 3 above)
Comment (gaocegege) [reply pkg/registry/serving/composer.go:364]: Master will be updated frequently, then we can use a commit to keep the link always pointing to the right code block.

Comment (simon-cj) [reply pkg/registry/serving/composer.go:364]: Got it, SGTM

File: pkg/registry/serving/composer.go [hunk 5]
```diff
@@ -273,14 +279,21 @@ func getModelFormat(pu *seldonv1.PredictiveUnit) string {
...
+		fallthrough
+	case string(modeljobsv1alpha1.FormatXGBoost):
 		return viper.GetString(envMLServerImage)
-	}
 
-	return viper.GetString(envTRTServingImage)
+	// Group3 for TRT server image
+	default:
+		return viper.GetString(envTRTServingImage)
+	}
```
Comment (simon-cj) [inline pkg/registry/serving/composer.go:296]: For MLflow, we need not set image, it will be set by seldon core? Is that right?

Comment (Thrimbda) [reply pkg/registry/serving/composer.go:418]: yes, it will be set by Seldon core

File: pkg/registry/serving/composer.go (hunk 4 above)
Comment (Thrimbda) [reply pkg/registry/serving/composer.go]: make sense

File: pkg/registry/serving/composer.go [hunk 6]
```diff
@@ -83,43 +83,38 @@ func validateComponentSpecs(p *seldonv1.PredictorSpec) error {
...
-		} else {
-			// For default image
-			err := composeDefaultUserContainer(sdep, &p.Graph, componentSpecMap)
-			if err != nil {
-				return err
+
+			componentSpecMap := getComponentsMap(p.ComponentSpecs)
+
+			// Compose user containers
+			if p.ComponentSpecs[0].Spec.Containers[0].Image != "" {
```
Comment (gaocegege) [inline pkg/registry/serving/composer.go:95]: 是不是应该 check 下数组长度？

File: pkg/registry/serving/composer.go [hunk 7]
```diff
@@ -472,7 +482,11 @@ func composeInitContainer(sdep *seldonv1.SeldonDeployment, pu *seldonv1.Predicto
 	modelMountPath := getModelMountPath(userContainer, sdep.Name)
 
 	initContainer := &corev1.Container{
-		Name:  "model-initializer",
+		// mimics the behavior of seldon model initializer for it will disable the default init container injection
+		// in case of a pre-packaged server implementation was selected: https://github.com/SeldonIO/seldon-core/blob/0bd83773228a18e7f376270f4b85cbef69395b8f/operator/controllers/model_initializer_injector.go#L142
+		// the default name will be generated here:
+		// https://github.com/SeldonIO/seldon-core/blob/0ef45fd234a674fc9b6c8d034cd2e42b4c9ebd05/operator/controllers/model_initializer_injector.go#L118
+		Name:  pu.Name + "-" + "model-initializer",
```
Comment (gaocegege) [inline pkg/registry/serving/composer.go:489]:
```suggestion
		Name:  pu.Name + "-model-initializer",
```

File: pkg/registry/serving/composer.go (hunk 6 above)
Comment (Thrimbda) [reply pkg/registry/serving/composer.go:95]: 这个地方在上面的validateComponentSpecs有检查

**Comment by simon-cj** [comment]
@Thrimbda CI is not pass

**Comment by Thrimbda** [comment]
> @Thrimbda CI is not pass

yes, I am holding it

File: pkg/registry/serving/composer.go [hunk 8]
```diff
@@ -106,13 +106,12 @@ func Compose(sdep *seldonv1.SeldonDeployment) error {
...
 
+			// Conpose init container for pod
+			composeInitContainer(sdep, &p)
 		}
 
 		// Setup no-engine mode
 		setupNoEngineMode(&p)
-
-		// Conpose init container for pod
-		composeInitContainer(sdep, &p)
```
Comment (Thrimbda) [inline pkg/registry/serving/composer.go:122]: 这个改动的原因是在 MLflow 的情况下完全使用 seldon core 预定义的 init container，因此不再需要自定义一个 init container 了

**Pull request merged**